core/src/spotify_uri.rs

```diff
@@ -0,0 +1,577 @@
+use crate::{Error, SpotifyId};
+use std::{borrow::Cow, fmt};
+use thiserror::Error;
+
+use librespot_protocol as protocol;
+
+const SPOTIFY_ITEM_TYPE_ALBUM: &str = "album";
+const SPOTIFY_ITEM_TYPE_ARTIST: &str = "artist";
+const SPOTIFY_ITEM_TYPE_EPISODE: &str = "episode";
+const SPOTIFY_ITEM_TYPE_PLAYLIST: &str = "playlist";
+const SPOTIFY_ITEM_TYPE_SHOW: &str = "show";
+const SPOTIFY_ITEM_TYPE_TRACK: &str = "track";
+const SPOTIFY_ITEM_TYPE_LOCAL: &str = "local";
+const SPOTIFY_ITEM_TYPE_UNKNOWN: &str = "unknown";
+
+#[derive(Debug, Error, Clone, Copy, PartialEq, Eq)]
+pub enum SpotifyUriError {
+    #[error("not a valid Spotify URI")]
+    InvalidFormat,
+    #[error("URI does not belong to Spotify")]
+    InvalidRoot,
+}
+
+impl From<SpotifyUriError> for Error {
+    fn from(err: SpotifyUriError) -> Self {
+        Error::invalid_argument(err)
+    }
+}
+
+pub type SpotifyUriResult = Result<SpotifyUri, Error>;
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub enum SpotifyUri {
+    Album {
+        id: SpotifyId,
+    },
+    Artist {
+        id: SpotifyId,
+    },
+    Episode {
+        id: SpotifyId,
+    },
+    Playlist {
+        user: Option<String>,
+        id: SpotifyId,
+    },
+    Show {
+        id: SpotifyId,
+    },
+    Track {
+        id: SpotifyId,
+    },
+    Local {
+        artist: String,
+        album_title: String,
+        track_title: String,
+        duration: std::time::Duration,
+    },
+    Unknown {
+        kind: Cow<'static, str>,
+        id: String,
+    },
+}
+
+impl SpotifyUri {
+    /// Returns whether this `SpotifyUri` is for a playable audio item, if known.
+    pub fn is_playable(&self) -> bool {
+        matches!(self, SpotifyUri::Episode { .. } | SpotifyUri::Track { .. })
+    }
+
+    /// Gets the item type of this URI as a static string
+    pub fn item_type(&self) -> &'static str {
+        match &self {
+            SpotifyUri::Album { .. } => SPOTIFY_ITEM_TYPE_ALBUM,
+            SpotifyUri::Artist { .. } => SPOTIFY_ITEM_TYPE_ARTIST,
+            SpotifyUri::Episode { .. } => SPOTIFY_ITEM_TYPE_EPISODE,
+            SpotifyUri::Playlist { .. } => SPOTIFY_ITEM_TYPE_PLAYLIST,
+            SpotifyUri::Show { .. } => SPOTIFY_ITEM_TYPE_SHOW,
+            SpotifyUri::Track { .. } => SPOTIFY_ITEM_TYPE_TRACK,
+            SpotifyUri::Local { .. } => SPOTIFY_ITEM_TYPE_LOCAL,
+            SpotifyUri::Unknown { .. } => SPOTIFY_ITEM_TYPE_UNKNOWN,
+        }
+    }
+
+    /// Gets the ID of this URI. The resource ID is the component of the URI that identifies
+    /// the resource after its type label. If `self` is a named ID, the user will be omitted.
+    pub fn to_id(&self) -> Result<String, Error> {
```

Comment:
Was just looking through at some code and I don't think there is any reason for us to return `Result<String, Error>` here instead of just `String` as we never appear to return `Err`... should we keep it in case we need to later or not?